tests/0122-ssl_cb.c

```diff
@@ -0,0 +1,308 @@
+    }
+    rd_kafka_metadata_destroy(md);
+    rd_kafka_destroy(rk);
+    return;
+}
+
+
+static void read_ssl_file_from_env(const char *envvar, char **dest, size_t *length) {
+    const char *path_val;
+    size_t szrt;
+    FILE *fh;
+
+    path_val = test_getenv(envvar, NULL);
+    if (!path_val) {
+        TEST_SKIP("Test requires %s env var\n", envvar);
+        return;
+    }
+    fh = fopen(path_val, "rb");
+    if (!fh) {
+        TEST_FAIL_LATER("File %s could not be opened: %m\n", path_val);
+        return;
+    }
+    fseek(fh, 0, SEEK_END);
+    *length = ftell(fh);
+    fseek(fh, 0, SEEK_SET);
+
+    *dest = malloc(*length);
+    // should probably handle short reads, but *shrug*
+    szrt = fread(*dest, 1, *length, fh);
+    if (!szrt) {
+        TEST_FAIL_LATER("File %s could not be read: %m\n", path_val);
+        fclose(fh);
+        return;
+    }
+    fclose(fh);
+    return;
+}
+
+int main_0122_ssl_cb(int argc, char **argv) {
+    struct test_context ctx;
+
+    memset(&ctx, 0, sizeof(struct test_context));
+
+    if (!test_check_builtin("ssl")) {
+        TEST_SKIP("Test requires SSL support\n");
+        goto cleanup;
+    }
+    if (!test_check_builtin("ssl_client_cert_callback")) {
+        TEST_SKIP("Test requires ssl_client_cert_callback support (i.e. built with openssl >= 1.0.2\n");
+        goto cleanup;
+    }
+
+
+    read_ssl_file_from_env(
+            "RDK_SSL_ca_pem", &ctx.ca_cert_pem, &ctx.ca_cert_pem_length
+    );
+    read_ssl_file_from_env(
+            "RDK_SSL_pub_pem", &ctx.direct_signed_client_cert_pem, &ctx.direct_signed_client_cert_pem_length
+    );
+    read_ssl_file_from_env(
+            "RDK_SSL_priv_pem", &ctx.direct_signed_client_key_pem, &ctx.direct_signed_client_key_pem_length
+    );
+    read_ssl_file_from_env(
+            "RDK_UNTRUSTEDSSL_pub_pem", &ctx.self_signed_client_cert_pem, &ctx.self_signed_client_cert_pem_length
+    );
+    read_ssl_file_from_env(
+            "RDK_UNTRUSTEDSSL_priv_pem", &ctx.self_signed_client_key_pem, &ctx.self_signed_client_key_pem_length
+    );
+    read_ssl_file_from_env(
+            "RDK_INTERMEDIATESSL_pub_pem", &ctx.intermediate_signed_client_cert_pem, &ctx.intermediate_signed_client_cert_pem_length
+    );
+    read_ssl_file_from_env(
+            "RDK_INTERMEDIATESSL_priv_pem", &ctx.intermediate_signed_client_key_pem, &ctx.intermediate_signed_client_key_pem_length
+    );
+    read_ssl_file_from_env(
+            "RDK_INTERMEDIATESSL_intermediate_pub_pem", &ctx.intermediate_signed_intermediate_cert_pem, &ctx.intermediate_signed_intermediate_cert_pem_length
+    );
+    read_ssl_file_from_env(
+            "RDK_INTERMEDIATESSL_intermediate_priv_pem", &ctx.intermediate_signed_intermediate_key_pem, &ctx.intermediate_signed_intermediate_key_pem_length
+    );
+    if (test_curr->state == TEST_SKIPPED || test_curr->state == TEST_FAILED) {
+        goto cleanup;
+    }
+
+    do_test_ssl_cb(&ctx, 1, ssl_client_cert_cb_good_direct);
```

Comment:
This is not really a callback, so rename it to do_test_ssl_cert() and also pass it a `const char *what` string that says what the test is doing.
Then in do_test_ssl_cert() before any of the code you do:
```c
SUB_TEST_QUICK("%s", what);
....all the testing...
SUB_TEST_PASS();
}
```

This will make it easier to track down which test is failing, and allow you to run a specific subtest with:
`TESTS=0124 SUBTESTS=intermediate make`